### ISSUE TITLE
Chaturbate - constant buffering ever few seconds on almost all cams

### ISSUE THREAD

**Issue by wcuser** [whitecream01/WhiteCream-V0.0.1#35]
Setup: Nvidia Shield / Jarvis, latest whitecream (.42). All network connections are ethernet, no wireless. up/down are 150/150, speedtest.net shows like 5ms latency locally, and only 80ms latency to West Coast, but nearly full throughput the whole way. I doubt its my connection. 

I think something has changed recently with the Chaturbate streaming. Prior to some of the more recent updates, I was able to watch various cams with no buffering or even dropped frames at all, but now, every cam drops a lot of frames, constantly stutters every 2-3 seconds, and always runs out the buffer. Needless to say, it makes the app sorta useless. 

Now, to isolate this, I did the following: 
1. Tested Kodi with no stream buffering in advancedsettings.xml - drops frames, buffers, stutters
2. Tested Kodi with stream buffering in advancedsettings.xml - still drops frames, buffers, stutters
3. Tested Kodi on Windows 10 with/without stream buffering in advset.xml - still drops frames, buffers, stutters
4. Test chaturbate from Chrome on Windows 10 - no issues whatsoever - I believe this isolates any possible networking issue from being the cause, it works fine in a browser on the same machine Kodi buffers/stutters/drops frames. 
5. Using "Livestreamer" on linux also doesn't suffer from any buffering or stuttering, can't tell if it drops any frames or not, doesn't look like it. 

Basically, everything I've tried narrows the stuttering/buffering/droppedframes down to the WC app as best I can manage. Any other method for viewing seems to work just fine from where I'm at. This wasn't always the case either, it used to work just fine with no stuttering/buffering. Also, you can't seem to change pages anymore either. 

Does the WC app use its own internal buffering, or does it use the Kodi system wide buffer settings? If internal, can you point out where they are set so I can see if tweaking them helps at all? Or, if it can just plain be fixed, that would be cool too. 

Love porn, love the app.

**Comment by hdgdl** [comment]
WC only provides the hls stream link (http://.......m3u8) to kodi.
Playback is done by kodi...like in every other addon too (twitch, youtube etc)

i would suggest to create a thread on kodi forums with debuglog while playing the stream and see if any dev will pick it up

**Comment by whitecream01** [comment]
It's not just Kodi, it's the m3u8 streams.
For example, go to their website, open a cam, check the source, locate the m3u8 file and paste that in: https://cdnsun.com/knowledgebase/tools/test-player
Click 'test', and then play it with the jwplayer. That one plays bad too.

I looked at using the rtmp streams (that plays fine) in Kodi, but that doesn't work. The librtmp shipped with Kodi doesn't support the commands needed for chaturbate (CheckPublicStatus).

**Comment by wcuser** [comment]
wc01: Thanks, that does make sense why Chrome/Livestreamer don't suffer from that I suppose. 

hdgdl: Kodi buffering in advancedsettings.xml is a readahead buffer I guess, which works fine if you have a known data set like a static file, but readahead won't work on any live stream because, well, its live and there is no future data to readahead on. I suppose the only way they could fix that is if they adopted another buffer method to delay playback a few seconds to let the buffer fill to an acceptable level first. I've seen how the Kodi dev's treat plebs, I'm not going to suggest it. Never seen a more popular piece of software with more assholes at the helm ever. 

Thanks for the responses anyways, if this isn't something you can fix, then its not something you can fix and close this out I guess.

**Comment by hdgdl** [comment]
i tried all day to get rtmp running....even with the new inputstream.rtmp addon
After many frustating hours i gave up...on windows you dont get any amf StuffIt going with inputstream you only get amf1 but we need amf3.
Maybe next weekend i get in to it again and try to reason with some devs
Got the same probs with Naked.com flash streams and ifriends.net

**Comment by wcuser** [comment]
Well, on the off chance no one thanks you, Thank You. Yer doing gods work my man. ;)

**Comment by whitecream01** [comment]
hdgdl, librtmp needs to patched with this for chaturbate to work: https://github.com/aajanki/rtmpdump/blob/e3cde4e4cd2d0f86e261c89e98e17e7b9d43165c/librtmp/rtmp.c#L2492

Like this: https://github.com/quarnster/boxeebox-xbmc/blob/master/tools/depends/target/librtmp/librtmp-60-second-fix.patch :)

**Comment by mortael** [comment]
Ah, there is a precompiled librtmp with the patch here: http://kodi.wiki/view/how-to:update_librtmp

**Comment by whitecream01** [comment]
Ah, there is a precompiled librtmp with the patch here: http://kodi.wiki/view/how-to:update_librtmp

**Comment by whitecream01** [comment]
I got the rtmp streams working (pre-Krypton) with a patched librtmp from the Kodi wiki.
Only, no audio....the audio stream is not supported.

**Comment by wcuser** [comment]
Thanks for continuing to look into this, I appreciate it. 

If librtmp hacking yields no results, how about something like this for the m3u8 files only: 

http://kodi.wiki/view/HOW-TO:Use_external_players_on_Android

I realize this isn't something you can control from the addon side, I'm not asking you to. Having someone have to install some external player is quite lame, but for some of us who aren't afraid of external players, do you see any issues with trying to kludge something like this up? (I'd be doing the kludging, not you, just looking for any "gotchas" you might see that I don't...) There are a LOT of external players supported on Android it looks like, and maybe something like launching VLC for m3u8 files would be a cheesy but functional workaround.

**Comment by wcuser** [comment]
So, I got external players working for filetype "m3u8" and it did launch an external player, but same choppy playback and instead of buffering, VLC just stops when it runs out of cache. Meh. I wonder if it uses an unpatched librtmp as well. 

So, if you figure out the whole librtmp business, even if it takes installing some extra stuff, I'm game to test it.

**Comment by whitecream01** [comment]
The m3u8 way and RTMP are different things.

The RTMP stream is played when you visit the cam in a browser (flash). That one plays fine.
The m3u8 stream is what I use in the addon, that one is bad. It's the stream that is bad, also with other players or in a browser (using jwplayer).

To get the RTMP stream to work with Kodi, it needs a patched librtmp. The streams plays like on their site, but without audio because Speex audio is not supported in Kodi.

I tried playing the m3u8 stream with the f4mproxy module in Kodi, that is very slightly better (most of the times). But still very annoying.

**Comment by wcuser** [comment]
I wonder what changed recently though. Lets say four weeks ago I could watch any number of the top cams on the site through your addon with absolutely no issue whatsoever, then sometime in the last month (I can't pinpoint it, its not like I watch it every day) the throughput just tanked into unwatchable. 

Anyways, thanks for looking into it, but like I said before, if its not something you can fix on the addon side, then there isn't anything you can fix on the addon side.

**Comment by whitecream01** [comment]
I just pushed a release where you can choose different options in the addon settings.
Direct (like previous versions)
F4mProxy
RTMP (needs a patched librtmp)

Mostly no audio with RTMP, i just had one stream while testing that had audio :)

**Comment by wcuser** [comment]
Any tips on a precompiled/patched librtmp for Android (arm)? 

I'll give the F4mProxy a shot when I get home. I assume the setup is just to install the F4mProxy from Fusion?

**Comment by mortael** [comment]
I just found this: https://www.mediafire.com/folder/3623pdo81n353/lib#3imvdytim8t3b

**Comment by wcuser** [comment]
Actually, WC provides a link right where you set it up in the addon, but that link above looks much better organized and has a lot of other platforms all in one spot, so that's a good find. 

I just messed around with this a bit, and wc01 is pretty much dead on in his analysis of it. 

- F4mProxy maybe works marginally better than straight rtmp, but still suffers from the same problem, at least on my end. Stuttering and choppy video. Any improvement with F4mProxy is negligible and probably not worth inclusion, at least from the Android users perspective. I didn't try it on *Elec or straight Linux. Maybe its better there. 

- Replacing librtmp from the links provided by the wc addon in setup yields a stable video playback, but no audio. So close, yet so far. I'm a bit surprised that Kodi doesn't support the Speex audio codec. It's listed in [http://kodi.wiki/view/Features_and_supported_formats](url) but maybe librtmp mangles it somehow? 

Again, thanks for the time spent on this.

**Comment by wcuser** [comment]
So, for laughs I compiled Kodi from scratch with the built in ffmpeg enabling libspeex and swapped in the updated librtmp.so fix on a ubuntu 16.04 box, and it sort of worked. I got stable playback and audio, but only mono audio. And, Kodi crashed when you stopped a stream. 

So, I think unless Chaturbate fixes it on their end, which I'd imagine they have zero motivation to do, might be SOL. 

I have one other idea, but it would only work on Linux/Win32 (no clue about Android) - I might make an external player hook for streamlink (an updated fork of livestreamer) which is a python player and see if that works. 

@whitecream01 : when using the librtmp config on your latest update, I'd assume that Kodi can redirect to an external player by hooking protocol type "rtmp", or if not, what is the new filetype if it isn't using the m3u8 filetype anymore? I want to try to redirect your librtmp stream that plays stable to an external player to try to bypass the audio problem all together.

**Comment by wcuser** [comment]
(Note: Livestreamer is different from the LiveStream kodi addon.)

I just don't have the python/kodi addon knowledge to make livestreamer work I don't think. 

If you are interested, livestreamer is a pure python app that will play lots of different flash streams, including chaturbate. It also has a module system so you can add on to it if you know how, so it might work with some of the other sites you said you were stuck with m3u8 streams. 

Apparently, it used to also be a default Kodi addon, but isn't anymore, but I can't figure out why, except that the official "livestreamer" project looks abandoned, but it has been forked (recently) to "streamlink" and that looks fairly active. 

Links (if you are interested): 

"Livestreamer" website (good docs)
[http://docs.livestreamer.io/](url)

"Livestreamer" github (looks idle)
[https://github.com/chrippa/livestreamer](url)

"Streamlink" github (fork of Livestreamer, activity)
[https://github.com/streamlink/streamlink](url)

Looks like it used to be an addon: 
[http://kodi.wiki/view/Add-on:Livestreamer](url)

Possible fork of the Livestreamer addon (not current)
[https://github.com/blxd/plugin.video.livestreamer](url)

**Comment by hdgdl** [comment]
anyone check the latest krypton nightly ?
Interesting pull request regarding hls
https://github.com/xbmc/xbmc/pull/10766

If i recall correctly from the m3u8 manifest, chaturbate uses 3 second chunks...so maybe the buffering comes from that jumpback to start..

**Comment by wcuser** [comment]
I haven't, but I might give that a shot this weekend. Good lookin out.

**Comment by wcuser** [comment]
Heh, actually set it to compile while I'm on my way home from work. 

Confirmed I was indeed compiling the updated code from #10766. 

No real change. This was on Ubuntu after compiling the source. It still stutters and audio drops every few seconds.

**Comment by whitecream01** [comment]
Livestreamer/streamlink just loads the m3u8 file from the site. It's really the chaturbate m3u8 stream that is bad, looks like they did it on purpose for us/others. So we would use the site.

**Comment by wcuser** [comment]
Yeah, I tried it with VLC and it still stutters all over the place. 

Anyways, again, thanks for lookin into it. You can close this out if you want.

**Comment by wcuser** [comment]
So I sent CB a support request saying their throughput on VLC was total shit and I think they actually adjusted it. Buffering is near non-existent on straight connect now.
